controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -275,6 +282,184 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
+                    e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, null)
+                    .thenCompose(subscribersData -> {
+                        //3. if SubscribersRecord does not exist return empty list
+                        if (subscribersData == null) {
+                            return CompletableFuture.completedFuture(emptySubscribersList);
+                        }
+                        return CompletableFuture.completedFuture(subscribersData.keySet().stream().collect(Collectors.toList()));
+                    }).exceptionally(ex -> {
+                              log.warn(requestId, "Exception thrown in trying to get list of Stream subscribers. {}",
+                                      ex.getMessage());
+                              Throwable cause = Exceptions.unwrap(ex);
+                              if (cause instanceof TimeoutException) {
+                                  throw new CompletionException(cause);
+                              } else {
+                                  log.warn(requestId, "getSubscribersForStream failed due to {}", cause);
+                                  return emptySubscribersList;
+                              }
+                          });
+                  }), e -> Exceptions.unwrap(e) instanceof RetryableException, SUBSCRIBER_OPERATION_RETRIES, executor);
+    }
+
+    /**
+     * Remove a subscriber from subscribers' metadata.
+     * Needed for Consumption based retention.
+     * @param scope      scope.
+     * @param stream     stream name.
+     * @param subscriber  Stream Subscriber publishing the StreamCut.
+     * @param truncationStreamCut  Truncation StreamCut.
+     * @param contextOpt optional context
+     * @return update status.
+     */
+    public CompletableFuture<UpdateSubscriberStatus.Status> updateTruncationStreamCut(String scope, String stream,
+                                                                             String subscriber, ImmutableMap<Long, Long> truncationStreamCut,
+                                                                             OperationContext contextOpt) {
+        final OperationContext context = contextOpt == null ? streamMetadataStore.createContext(scope, stream) : contextOpt;
+        final long requestId = requestTracker.getRequestIdFor("updateTruncationStreamCut", scope, stream);
+
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.checkStreamExists(scope, stream)
+                .thenCompose(exists -> {
+                    // 1. check Stream exists
+                    if (!exists) {
+                        return CompletableFuture.completedFuture(UpdateSubscriberStatus.Status.STREAM_NOT_FOUND);
+                    }
+                    // 2. check if StreamCut is valid
+                    return streamMetadataStore.isStreamCutValid(scope, stream, truncationStreamCut, context, executor)
```

Comment:
we should also make use of "greater" than property to see if the new streamcut is strictly greater than equal to previously reported truncation streamcut for the given subscriber so that it always move ahead.

Reply:
There may be "recovery" scenarios where a Reader Group restarts from an earlier Stream-Cut and hence publishes an older StreamCut as a truncation Stream-Cut later.... so I would not want to add that check here...

Reply:
but can stream cut be "reset" back to an older streamcut? 
what if controller had already used the more recent one for truncation.
we should strictly move the streamcuts forward..
it also takes care of scenarios where multiple different controllers received different requests and if newer update is already set, the stale value should not override it.
take for example:

T0: rg ->SC1 -> controller1
t1: connection failure
t2: rg ->SC2 -> controller2

controller1 is attempting to update SC1 to segment store and controller 2 is attempting to update SC2.. 

if SC2 is updated, we should reject SC1.

Reply:
The validation you are suggesting can be added at a later point, if needed.
Do we do this check with manual truncation API today?

Reply:
yes we do, we check that the streamcut is actually moving forward. 
for each range, if segment number in one streamcut is same or higher, then that stream cut is greater than.

Reply:
Added this validation.